File: frontend/lib/supplier-language.ts
export const SUPPLIER_LANGUAGE_OPTIONS = [
  { value: "en", label: "English" },
  { value: "ms", label: "Malay" },
  { value: "zh-CN", label: "Chinese (Simplified)" },
  { value: "zh-TW", label: "Chinese (Traditional)" },
  { value: "ta", label: "Tamil" },
  { value: "id", label: "Indonesian" },
  { value: "th", label: "Thai" },
  { value: "vi", label: "Vietnamese" },
  { value: "fil", label: "Filipino" },
  { value: "ja", label: "Japanese" },
  { value: "ko", label: "Korean" },
  { value: "hi", label: "Hindi" },
  { value: "bn", label: "Bengali" },
  { value: "ur", label: "Urdu" },
  { value: "ar", label: "Arabic" },
  { value: "tr", label: "Turkish" },
  { value: "fr", label: "French" },
  { value: "de", label: "German" },
  { value: "es", label: "Spanish" },
  { value: "pt", label: "Portuguese" },
  { value: "pt-BR", label: "Portuguese (Brazil)" },
  { value: "it", label: "Italian" },
  { value: "nl", label: "Dutch" },
  { value: "pl", label: "Polish" },
  { value: "ru", label: "Russian" },
  { value: "uk", label: "Ukrainian" },
  { value: "cs", label: "Czech" },
  { value: "ro", label: "Romanian" },
  { value: "el", label: "Greek" },
  { value: "he", label: "Hebrew" },
  { value: "fa", label: "Persian" },
  { value: "sw", label: "Swahili" },
  { value: "af", label: "Afrikaans" },
  { value: "my", label: "Burmese" },
  { value: "km", label: "Khmer" },
  { value: "lo", label: "Lao" },
  { value: "ne", label: "Nepali" },
  { value: "si", label: "Sinhala" },
  { value: "te", label: "Telugu" },
  { value: "ml", label: "Malayalam" },
  { value: "kn", label: "Kannada" },
  { value: "gu", label: "Gujarati" },
  { value: "pa", label: "Punjabi" },
  { value: "mr", label: "Marathi" },
  { value: "no", label: "Norwegian" },
  { value: "sv", label: "Swedish" },
  { value: "da", label: "Danish" },
  { value: "fi", label: "Finnish" },
] as const

export type SupplierPreferredLanguage = string
export type SupplierMessageLanguage = string

export const DEFAULT_SUPPLIER_LANGUAGE: SupplierPreferredLanguage = "en"

const LANGUAGE_BY_NORMALIZED_CODE = new Map(
  SUPPLIER_LANGUAGE_OPTIONS.map((option) => [
    option.value.toLowerCase(),
    option,
  ])
)

export function isSupplierPreferredLanguage(value: string) {
  return LANGUAGE_BY_NORMALIZED_CODE.has(value.trim().toLowerCase())
}

export function normalizeSupplierPreferredLanguage(
  value?: string | null
): SupplierPreferredLanguage {
  if (!value) return DEFAULT_SUPPLIER_LANGUAGE

  const normalized = value.trim().toLowerCase()
  return LANGUAGE_BY_NORMALIZED_CODE.get(normalized)?.value ?? DEFAULT_SUPPLIER_LANGUAGE
}

export function getLanguageLabel(code?: string | null) {
  const language = LANGUAGE_BY_NORMALIZED_CODE.get(
    normalizeSupplierPreferredLanguage(code).toLowerCase()
  )
  return language?.label ?? "English"
}

export function getMessageLanguageCode(code?: string | null): SupplierMessageLanguage {
  return normalizeSupplierPreferredLanguage(code)
}

export function getMessageLanguageBadge(code?: string | null) {
  const normalized = normalizeSupplierPreferredLanguage(code)
  if (normalized === "zh-CN") return { short: "中文", full: "Chinese (Simplified)" }
  if (normalized === "zh-TW") return { short: "繁中", full: "Chinese (Traditional)" }

  const label = getLanguageLabel(normalized)
  return {
    short: normalized.toUpperCase(),
    full: label,
  }
}

File: backend/app/ai/supplier_language.py
from __future__ import annotations

from typing import Final


DEFAULT_SUPPLIER_LANGUAGE: Final[str] = "en"
SUPPLIER_LANGUAGE_LABELS: Final[dict[str, str]] = {
    "en": "English",
    "ms": "Malay",
    "zh-CN": "Chinese (Simplified)",
    "zh-TW": "Chinese (Traditional)",
    "ta": "Tamil",
    "id": "Indonesian",
    "th": "Thai",
    "vi": "Vietnamese",
    "fil": "Filipino",
    "ja": "Japanese",
    "ko": "Korean",
    "hi": "Hindi",
    "bn": "Bengali",
    "ur": "Urdu",
    "ar": "Arabic",
    "tr": "Turkish",
    "fr": "French",
    "de": "German",
    "es": "Spanish",
    "pt": "Portuguese",
    "pt-BR": "Portuguese (Brazil)",
    "it": "Italian",
    "nl": "Dutch",
    "pl": "Polish",
    "ru": "Russian",
    "uk": "Ukrainian",
    "cs": "Czech",
    "ro": "Romanian",
    "el": "Greek",
    "he": "Hebrew",
    "fa": "Persian",
    "sw": "Swahili",
    "af": "Afrikaans",
    "my": "Burmese",
    "km": "Khmer",
    "lo": "Lao",
    "ne": "Nepali",
    "si": "Sinhala",
    "te": "Telugu",
    "ml": "Malayalam",
    "kn": "Kannada",
    "gu": "Gujarati",
    "pa": "Punjabi",
    "mr": "Marathi",
    "no": "Norwegian",
    "sv": "Swedish",
    "da": "Danish",
    "fi": "Finnish",
}

CANONICAL_LANGUAGE_CODES: Final[dict[str, str]] = {
    code.lower(): code for code in SUPPLIER_LANGUAGE_LABELS
}


def normalize_supplier_language(value: str | None) -> str:
    code = (value or "").strip()
    if not code:
        return DEFAULT_SUPPLIER_LANGUAGE

    return CANONICAL_LANGUAGE_CODES.get(code.lower(), DEFAULT_SUPPLIER_LANGUAGE)


def supplier_language_label(value: str | None) -> str:
    return SUPPLIER_LANGUAGE_LABELS[normalize_supplier_language(value)]
